### PR TITLE
TMA-1014: parallel do_not_touch logic in UFB

### PR DESCRIPTION
**Pull request by kubamahnert**

### CODE REVIEW

File: lib/gooddata/lcm/actions/synchronize_user_filters.rb [hunk 1]
```diff
@@ -167,27 +167,14 @@ def call(params)
...
-                  puts "Client #{c.client_id} has no project."
-                  next
-                end
-                if current_project.deleted?
-                  puts "Project #{current_project.pid} of client #{c.client_id} is deleted."
-                  next
-                end
-
-                puts "Delete all filters in project #{current_project.pid} of client #{c.client_id}"
+                current_project = c.project
```
Comment (kubamahnert) [inline lib/gooddata/lcm/actions/synchronize_user_filters.rb]: TODO: do not fail if something is wrong with the project

File: spec/lcm/integration/user_provisioning_end_to_end_spec.rb [hunk 2]
```diff
@@ -1,184 +1,57 @@
...
-      muf_complexity = 4
-      mufs = set_up_random_muf_data(@projects.first, @user_data, muf_complexity)
-      test_context = @test_context.merge(label: mufs.first[:label_id])
-      test_user_filters_brick(projects: @projects,
-                              test_context: test_context,
-                              mufs: mufs,
-                              muf_complexity: muf_complexity)
-    end
-  end
+      mufs = @fixtures[:clients].map do |client, project|
```
Comment (yenkins) [inline spec/lcm/integration/user_provisioning_end_to_end_spec.rb]: warning: Similar code found in user_provisioning_end_to_end_spec.rb:46, user_provisioning_extended_spec.rb:33 (mass = 44) (flay)

File: spec/lcm/integration/user_provisioning_extended_spec.rb [hunk 3]
```diff
@@ -0,0 +1,44 @@
...
+      builder_config = {
+        user_column: 'login',
+        labels: [{
+                   column: 'value',
+                   label: extra_mufs.first[:label_id]
+                 }]
+      }
+      extra_project.add_data_permissions GoodData::UserFilterBuilder.get_filters(GoodData::LCM2.convert_to_smart_hash(extra_mufs), GoodData::LCM2.convert_to_smart_hash(builder_config))
+
+      mufs = @project_fixtures[:clients].map do |client, project|
```
Comment (yenkins) [inline spec/lcm/integration/user_provisioning_extended_spec.rb]: warning: Similar code found in user_provisioning_end_to_end_spec.rb:46, user_provisioning_extended_spec.rb:33 (mass = 44) (flay)

File: lib/gooddata/lcm/actions/synchronize_user_filters.rb [hunk 4]
```diff
@@ -167,27 +167,14 @@ def call(params)
...
                 users = users_by_project[c.client_id]
-                delete_results = current_project.add_data_permissions([], run_params.merge(users_brick_input: users))
-                results.concat(delete_results[:results])
+                puts "Delete all filters in project #{current_project.pid} of client #{c.client_id}"
```
Comment (yenkins) [inline lib/gooddata/lcm/actions/synchronize_user_filters.rb]: info: Calls 'c.client_id' 3 times (DuplicateMethodCall) (reek)

File: lib/gooddata/lcm/actions/synchronize_user_filters.rb (hunk 1 above)
Comment (yenkins) [inline lib/gooddata/lcm/actions/synchronize_user_filters.rb]: info: Calls 'c.project' 2 times (DuplicateMethodCall) (reek)

File: lib/gooddata/lcm/actions/synchronize_user_filters.rb [hunk 5]
```diff
@@ -167,27 +167,14 @@ def call(params)
...
+                current_results = sync_user_filters(current_project, [], run_params.merge(users_brick_input: users), symbolized_config)
```
Comment (yenkins) [inline lib/gooddata/lcm/actions/synchronize_user_filters.rb]: info: Calls 'run_params.merge(users_brick_input: users)' 3 times (DuplicateMethodCall) (reek)

File: lib/gooddata/lcm/actions/synchronize_user_filters.rb [hunk 6]
```diff
@@ -167,27 +167,14 @@ def call(params)
             end
 
             unless run_params[:do_not_touch_filters_that_are_not_mentioned]
-              domain_clients.each do |c|
+              domain_clients.peach do |c|
```
Comment (yenkins) [inline lib/gooddata/lcm/actions/synchronize_user_filters.rb]: info: Has the variable name 'c' (UncommunicativeVariableName) (reek)

File: spec/lcm/integration/fixtures/base_fixtures.rb [hunk 7]
```diff
@@ -0,0 +1,13 @@
+module Fixtures
+  class BaseFixtures
```
Comment (yenkins) [inline spec/lcm/integration/fixtures/base_fixtures.rb]: info: Assumes too much for instance variable '@objects' (InstanceVariableAssumption) (reek)

Comment (yenkins) [inline spec/lcm/integration/fixtures/base_fixtures.rb]: info: Assumes too much for instance variable '@teardown' (InstanceVariableAssumption) (reek)

Comment (yenkins) [inline spec/lcm/integration/fixtures/base_fixtures.rb]: info: Has no descriptive comment (IrresponsibleModule) (reek)

File: spec/lcm/integration/fixtures/project_fixtures.rb [hunk 8]
```diff
@@ -0,0 +1,50 @@
...
+
+module Fixtures
+  class ProjectFixtures < BaseFixtures
+    def initialize(opts = {})
+      domain = opts[:domain]
+      rest_client = opts[:rest_client]
+      project_amount = opts[:project_amount] || 1
+      suffix = ConfigurationHelper.suffix
+
+      data_product = domain.create_data_product(id: FIXTURE_ID_PREFIX + suffix)
```
Comment (yenkins) [inline spec/lcm/integration/fixtures/project_fixtures.rb]: info: Calls 'fixture_id_prefix + suffix' 3 times (DuplicateMethodCall) (reek)

File: spec/lcm/integration/fixtures/project_fixtures.rb [hunk 9]
```diff
@@ -0,0 +1,50 @@
...
+      master_project = rest_client.create_project(title: FIXTURE_ID_PREFIX + suffix, auth_token: LcmConnectionHelper.environment[:prod_token])
```
Comment (yenkins) [inline spec/lcm/integration/fixtures/project_fixtures.rb]: info: Calls 'lcmconnectionhelper.environment' 2 times (DuplicateMethodCall) (reek)

Comment (yenkins) [inline spec/lcm/integration/fixtures/project_fixtures.rb]: info: Calls 'lcmconnectionhelper.environment[:prod_token]' 2 times (DuplicateMethodCall) (reek)

File: spec/lcm/integration/fixtures/project_fixtures.rb [hunk 10]
```diff
@@ -0,0 +1,50 @@
...
+          client: rest_client,
+          title: [FIXTURE_ID_PREFIX, suffix, n.to_s].join(' '),
+          auth_token: LcmConnectionHelper.environment[:prod_token],
+          environment: 'TESTING'
+        }
+        project_helper = Support::ProjectHelper.create(opts)
+        puts 'creating LDM and stuff'
+        project_helper.create_ldm
+        project_helper.load_data
+        puts "created project #{project_helper.project.pid}"
```
Comment (yenkins) [inline spec/lcm/integration/fixtures/project_fixtures.rb]: info: Calls 'project_helper.project' 2 times (DuplicateMethodCall) (reek)

File: spec/lcm/integration/fixtures/project_fixtures.rb [hunk 11]
```diff
@@ -0,0 +1,50 @@
+require_relative 'base_fixtures'
+require_relative '../support/configuration_helper'
+require_relative '../support/connection_helper'
+require_relative '../support/project_helper'
+
+module Fixtures
+  class ProjectFixtures < BaseFixtures
```
Comment (yenkins) [inline spec/lcm/integration/fixtures/project_fixtures.rb]: info: Has no descriptive comment (IrresponsibleModule) (reek)

File: spec/lcm/integration/fixtures/project_fixtures.rb [hunk 12]
```diff
@@ -0,0 +1,50 @@
...
+        puts "created project #{project_helper.project.pid}"
+        project_helper.project
+      end
+
+      @teardown = lambda do
+        (projects + [master_project]).map(&:delete)
+        data_product.delete(force: true)
+      end
+
+      clients = Hash[projects.each_with_index.map do |project, i|
```
Comment (yenkins) [inline spec/lcm/integration/fixtures/project_fixtures.rb]: info: Has the variable name 'i' (UncommunicativeVariableName) (reek)

File: spec/lcm/integration/fixtures/project_fixtures.rb [hunk 13]
```diff
@@ -0,0 +1,50 @@
...
+      domain = opts[:domain]
+      rest_client = opts[:rest_client]
+      project_amount = opts[:project_amount] || 1
+      suffix = ConfigurationHelper.suffix
+
+      data_product = domain.create_data_product(id: FIXTURE_ID_PREFIX + suffix)
+      master_project = rest_client.create_project(title: FIXTURE_ID_PREFIX + suffix, auth_token: LcmConnectionHelper.environment[:prod_token])
+      segment = data_product.create_segment(segment_id: FIXTURE_ID_PREFIX + suffix, master_project: master_project)
+
+      projects = project_amount.times.map do |n|
```
Comment (yenkins) [inline spec/lcm/integration/fixtures/project_fixtures.rb]: info: Has the variable name 'n' (UncommunicativeVariableName) (reek)

File: spec/lcm/integration/fixtures/user_provisioning_fixtures.rb [hunk 14]
```diff
@@ -0,0 +1,76 @@
...
+require_relative 'project_fixtures'
+
+module Fixtures
+  class UserProvisioningFixtures < BaseFixtures
+    FIXTURE_ID_PREFIX += 'userprov-e2e-'
+
+    def initialize(opts = {})
+      user_amount = opts[:user_amount] || 2
+      rest_client = LcmConnectionHelper.production_server_connection
+      domain = rest_client.domain(LcmConnectionHelper.environment[:prod_organization])
```
Comment (yenkins) [inline spec/lcm/integration/fixtures/user_provisioning_fixtures.rb]: info: Calls 'lcmconnectionhelper.environment' 2 times (DuplicateMethodCall) (reek)

File: spec/lcm/integration/fixtures/user_provisioning_fixtures.rb [hunk 15]
```diff
@@ -0,0 +1,76 @@
...
+      users = Array.new(user_amount) do |n|
+        {
+          custom_login: "lcm-userprov-e2e-UPPERCASE-#{n}@gooddata.com",
+          position: 'developer'
+        }
+      end
+
+      project_fixtures = ProjectFixtures.new domain: domain, rest_client: rest_client, project_amount: 2
+
+      user_data = project_fixtures[:clients].map do |c, p|
```
Comment (yenkins) [inline spec/lcm/integration/fixtures/user_provisioning_fixtures.rb]: info: Calls 'project_fixtures[:clients]' 3 times (DuplicateMethodCall) (reek)

Comment (yenkins) [inline spec/lcm/integration/fixtures/user_provisioning_fixtures.rb]: info: Calls 'project_fixtures[:clients].map' 2 times (DuplicateMethodCall) (reek)

Comment (yenkins) [inline spec/lcm/integration/fixtures/user_provisioning_fixtures.rb]: info: Refers to 'project_fixtures' more than self (maybe move it to another class?) (FeatureEnvy) (reek)

Comment (yenkins) [inline spec/lcm/integration/fixtures/user_provisioning_fixtures.rb]: info: Has the variable name 'c' (UncommunicativeVariableName) (reek)

Comment (yenkins) [inline spec/lcm/integration/fixtures/user_provisioning_fixtures.rb]: info: Has the variable name 'p' (UncommunicativeVariableName) (reek)

File: spec/lcm/integration/fixtures/user_provisioning_fixtures.rb [hunk 16]
```diff
@@ -0,0 +1,76 @@
...
+        domain.add_user(login: u[:custom_login]) unless domain.users u[:custom_login]
+      end
+
+      mufs = project_fixtures[:clients].map do |client, project|
+        Support::UserProvisioningHelper.random_muf_data client: client, project: project, users: users
+      end.flatten
+      Support::S3Helper.upload_file(ConfigurationHelper.csv_from_hashes(mufs), Support::S3Helper::USER_FILTERS_KEY)
+
+      brick_params = {
+        project_id: project_fixtures[:projects].first.pid, # this doesn't really matter for the runtime
```
Comment (yenkins) [inline spec/lcm/integration/fixtures/user_provisioning_fixtures.rb]: info: Calls 'project_fixtures[:projects]' 2 times (DuplicateMethodCall) (reek)

File: spec/lcm/integration/fixtures/user_provisioning_fixtures.rb [hunk 17]
```diff
@@ -0,0 +1,76 @@
...
+      user_data = project_fixtures[:clients].map do |c, p|
+        users.map do |u|
+          u.merge client_id: c.client_id,
+                  project_id: p.pid
+        end
+      end.flatten
+
+      Support::S3Helper.upload_file(ConfigurationHelper.csv_from_hashes(user_data), Support::S3Helper::USERS_KEY)
+      users.map do |u|
+        domain.add_user(login: u[:custom_login]) unless domain.users u[:custom_login]
```
Comment (yenkins) [inline spec/lcm/integration/fixtures/user_provisioning_fixtures.rb]: info: Calls 'u[:custom_login]' 2 times (DuplicateMethodCall) (reek)

File: spec/lcm/integration/fixtures/user_provisioning_fixtures.rb [hunk 18]
```diff
@@ -0,0 +1,76 @@
...
+        {
+          custom_login: "lcm-userprov-e2e-UPPERCASE-#{n}@gooddata.com",
+          position: 'developer'
+        }
+      end
+
+      project_fixtures = ProjectFixtures.new domain: domain, rest_client: rest_client, project_amount: 2
+
+      user_data = project_fixtures[:clients].map do |c, p|
+        users.map do |u|
```
Comment (yenkins) [inline spec/lcm/integration/fixtures/user_provisioning_fixtures.rb]: info: Calls 'users.map' 2 times (DuplicateMethodCall) (reek)

Comment (yenkins) [inline spec/lcm/integration/fixtures/user_provisioning_fixtures.rb]: info: Contains iterators nested 2 deep (NestedIterators) (reek)

Comment (yenkins) [inline spec/lcm/integration/fixtures/user_provisioning_fixtures.rb]: info: Has the variable name 'u' (UncommunicativeVariableName) (reek)

File: spec/lcm/integration/fixtures/user_provisioning_fixtures.rb [hunk 19]
```diff
@@ -0,0 +1,76 @@
...
+require_relative '../support/configuration_helper'
+require_relative '../support/connection_helper'
+require_relative '../support/s3_helper'
+require_relative '../support/lcm_helper'
+require_relative '../support/project_helper'
+require_relative 'base_fixtures'
+require_relative 'project_fixtures'
+
+module Fixtures
+  class UserProvisioningFixtures < BaseFixtures
```
Comment (yenkins) [inline spec/lcm/integration/fixtures/user_provisioning_fixtures.rb]: info: Has no descriptive comment (IrresponsibleModule) (reek)

File: spec/lcm/integration/fixtures/user_provisioning_fixtures.rb [hunk 20]
```diff
@@ -0,0 +1,76 @@
...
+    FIXTURE_ID_PREFIX += 'userprov-e2e-'
+
+    def initialize(opts = {})
+      user_amount = opts[:user_amount] || 2
+      rest_client = LcmConnectionHelper.production_server_connection
+      domain = rest_client.domain(LcmConnectionHelper.environment[:prod_organization])
+
+      users = Array.new(user_amount) do |n|
```
Comment (yenkins) [inline spec/lcm/integration/fixtures/user_provisioning_fixtures.rb]: info: Has the variable name 'n' (UncommunicativeVariableName) (reek)

File: spec/lcm/integration/support/user_provisioning_helper.rb [hunk 21]
```diff
@@ -0,0 +1,83 @@
...
+
+module Support
+  class UserProvisioningHelper
+    class << self
+      def uniq_label_with_value(project, value_count)
+        blacklists = $generated_uniq_labels_with_value || {}
+        blacklist = blacklists[project.pid] || []
+
+        atts = project.attributes.to_a.sort_by(&:identifier)
+        attribute = atts.shift while !attribute || attribute.date_attribute? || !attribute.primary_label || (blacklist.include? attribute)
```
Comment (yenkins) [inline spec/lcm/integration/support/user_provisioning_helper.rb]: info: Calls 'attribute.primary_label' 2 times (DuplicateMethodCall) (reek)

File: spec/lcm/integration/support/user_provisioning_helper.rb [hunk 22]
```diff
@@ -0,0 +1,83 @@
+require_relative '../brick_runner'
+
+module Support
+  class UserProvisioningHelper
+    class << self
+      def uniq_label_with_value(project, value_count)
+        blacklists = $generated_uniq_labels_with_value || {}
+        blacklist = blacklists[project.pid] || []
```
Comment (yenkins) [inline spec/lcm/integration/support/user_provisioning_helper.rb]: info: Calls 'project.pid' 2 times (DuplicateMethodCall) (reek)

File: spec/lcm/integration/support/user_provisioning_helper.rb [hunk 23]
```diff
@@ -0,0 +1,83 @@
...
+
+        atts = project.attributes.to_a.sort_by(&:identifier)
+        attribute = atts.shift while !attribute || attribute.date_attribute? || !attribute.primary_label || (blacklist.include? attribute)
+        label = attribute.primary_label
+        sorted_values = label.values.to_a.sort_by { |a| a[:value] }
+        notempty = sorted_values.reject { |v| v[:value].empty? }
```
Comment (yenkins) [inline spec/lcm/integration/support/user_provisioning_helper.rb]: info: Calls 'v[:value]' 2 times (DuplicateMethodCall) (reek)

Comment (yenkins) [inline spec/lcm/integration/support/user_provisioning_helper.rb]: info: Has the variable name 'v' (UncommunicativeVariableName) (reek)

File: spec/lcm/integration/support/user_provisioning_helper.rb [hunk 24]
```diff
@@ -0,0 +1,83 @@
+require_relative '../brick_runner'
+
+module Support
+  class UserProvisioningHelper
```
Comment (yenkins) [inline spec/lcm/integration/support/user_provisioning_helper.rb]: info: Has no descriptive comment (IrresponsibleModule) (reek)

File: spec/lcm/integration/support/user_provisioning_helper.rb [hunk 25]
```diff
@@ -0,0 +1,83 @@
...
+        client = opts[:client]
+        project = opts[:project]
+        users = opts[:users]
+        muf_complexity = opts[:muf_complexity] || 2
+
+        label, values = uniq_label_with_value project, muf_complexity
+        puts "Generated MUF for values #{values} of attribute #{label.title}"
+
+        users.map do |u|
+          values.map do |v|
```
Comment (yenkins) [inline spec/lcm/integration/support/user_provisioning_helper.rb]: info: Contains iterators nested 2 deep (NestedIterators) (reek)

Comment (yenkins) [inline spec/lcm/integration/support/user_provisioning_helper.rb]: info: Has the variable name 'v' (UncommunicativeVariableName) (reek)

File: spec/lcm/integration/user_provisioning_extended_spec.rb [hunk 26]
```diff
@@ -0,0 +1,44 @@
...
+    @project_fixtures.teardown
+  end
+
+  describe 'when provisioning in the declarative mode' do
+    it 'removes MUFs from projects not mentioned in the input' do
+      extra_client, extra_project = @project_fixtures[:clients].first
+      extra_mufs = Support::UserProvisioningHelper.random_muf_data client: extra_client, project: extra_project, users: @fixtures[:users]
+      builder_config = {
+        user_column: 'login',
+        labels: [{
```
Comment (kubamahnert) [inline spec/lcm/integration/user_provisioning_extended_spec.rb]: TODO: this should be declared only once and then reused!

**Comment by gdgate** [comment]
Build failed (check pipeline).

- https://checklist.intgdc.com/job/compliance/job/gooddata-ruby-opensource-repository-check/743/ : SUCCESS in 13s
- https://ci.intgdc.com/job/gooddata-ruby-pronto-code-review-zuul/1566/ : FAILURE in 1m 50s
- https://ci.intgdc.com/job/gooddata-ruby-unit-tests-zuul/1827/ : FAILURE in 1m 29s
- https://jenkins-ii.intgdc.com/job/gooddata-ruby/job/gooddata-ruby-check-pipeline/59/ : SUCCESS in 5m 57s

**Comment by gdgate** [comment]
Build failed (check pipeline).

- https://checklist.intgdc.com/job/compliance/job/gooddata-ruby-opensource-repository-check/744/ : SUCCESS in 9s
- https://ci.intgdc.com/job/gooddata-ruby-pronto-code-review-zuul/1567/ : FAILURE in 1m 42s
- https://ci.intgdc.com/job/gooddata-ruby-unit-tests-zuul/1828/ : FAILURE in 1m 46s
- https://jenkins-ii.intgdc.com/job/gooddata-ruby/job/gooddata-ruby-check-pipeline/60/ : SUCCESS in 6m 01s

File: spec/lcm/integration/user_provisioning_end_to_end_spec.rb (hunk 2 above)
Comment (yenkins) [inline spec/lcm/integration/user_provisioning_end_to_end_spec.rb]: warning: Similar code found in user_provisioning_end_to_end_spec.rb:46, user_provisioning_extended_spec.rb:36 (mass = 44) (flay)

File: spec/lcm/integration/user_provisioning_extended_spec.rb [hunk 27]
```diff
@@ -0,0 +1,47 @@
...
+          {
+            column: 'value',
+            label: extra_mufs.first[:label_id]
+          }
+        ]
+      }
+      extra_project.add_data_permissions GoodData::UserFilterBuilder.get_filters(GoodData::LCM2.convert_to_smart_hash(extra_mufs),
+                                                                                 GoodData::LCM2.convert_to_smart_hash(builder_config))
+
+      mufs = @project_fixtures[:clients].map do |client, project|
```
Comment (yenkins) [inline spec/lcm/integration/user_provisioning_extended_spec.rb]: warning: Similar code found in user_provisioning_end_to_end_spec.rb:46, user_provisioning_extended_spec.rb:36 (mass = 44) (flay)

File: lib/gooddata/lcm/actions/synchronize_user_filters.rb [hunk 28]
```diff
@@ -154,40 +158,34 @@ def call(params)
...
-                end
-                unless current_project
-                  puts "Client #{c.client_id} has no project."
-                  next
-                end
-                if current_project.deleted?
-                  puts "Project #{current_project.pid} of client #{c.client_id} is deleted."
-                  next
-                end
+                  users = users_by_project[c.client_id]
```
Comment (yenkins) [inline lib/gooddata/lcm/actions/synchronize_user_filters.rb]: info: Calls 'c.client_id' 4 times (DuplicateMethodCall) (reek)

File: lib/gooddata/lcm/actions/synchronize_user_filters.rb [hunk 29]
```diff
@@ -154,40 +158,34 @@ def call(params)
...
+                  puts "Delete all filters in project #{current_project.pid} of client #{c.client_id}"
+                  current_results = sync_user_filters(current_project, [], run_params.merge(users_brick_input: users), symbolized_config)
```
Comment (yenkins) [inline lib/gooddata/lcm/actions/synchronize_user_filters.rb]: info: Calls 'run_params.merge(users_brick_input: users)' 3 times (DuplicateMethodCall) (reek)

File: lib/gooddata/lcm/actions/synchronize_user_filters.rb [hunk 30]
```diff
@@ -154,40 +158,34 @@ def call(params)
...
 
-                puts "Delete all filters in project #{current_project.pid} of client #{c.client_id}"
-                users = users_by_project[c.client_id]
-                delete_results = current_project.add_data_permissions([], run_params.merge(users_brick_input: users))
-                results.concat(delete_results[:results])
+                  results.concat(current_results[:results])
+                rescue StandardError => e
```
Comment (yenkins) [inline lib/gooddata/lcm/actions/synchronize_user_filters.rb]: info: Has the variable name 'e' (UncommunicativeVariableName) (reek)

File: spec/lcm/integration/support/user_provisioning_helper.rb [hunk 31]
```diff
@@ -0,0 +1,83 @@
...
+        test_context = options[:test_context]
+        mufs = options[:mufs]
+        projects = options[:projects]
+
+        BrickRunner.user_filters_brick context: test_context, template_path: '../params/user_filters_brick_e2e.json.erb'
+
+        projects.each do |p|
+          filters = p.user_filters.to_a
+          label = p.attributes(mufs.first[:label_id])
+          expected_filters_by_email = mufs.select { |m| m[:project_id] == p.pid }.group_by { |f| f[:login] }
```
Comment (yenkins) [inline spec/lcm/integration/support/user_provisioning_helper.rb]: info: Contains iterators nested 2 deep (NestedIterators) (reek)

Comment (yenkins) [inline spec/lcm/integration/support/user_provisioning_helper.rb]: info: Has the variable name 'f' (UncommunicativeVariableName) (reek)

Comment (yenkins) [inline spec/lcm/integration/support/user_provisioning_helper.rb]: info: Has the variable name 'm' (UncommunicativeVariableName) (reek)

File: spec/lcm/integration/support/user_provisioning_helper.rb [hunk 32]
```diff
@@ -0,0 +1,83 @@
...
+
+          spec_env { expect(filters.length).to eq(expected_filters_by_email.length) }
+          expected_filters_by_email.each do |email, expected_filters|
+            # all emails are downcased during brick provisioning
+            matching = filters.find { |f| f.related.email == email.downcase }
```
Comment (yenkins) [inline spec/lcm/integration/support/user_provisioning_helper.rb]: info: Contains iterators nested 3 deep (NestedIterators) (reek)

File: spec/lcm/integration/support/user_provisioning_helper.rb [hunk 33]
```diff
@@ -0,0 +1,83 @@
...
+        sorted_values = label.values.to_a.sort_by { |a| a[:value] }
+        notempty = sorted_values.reject { |v| v[:value].empty? }
+        values = notempty.slice(0, value_count)
+
+        blacklists[project.pid] = blacklist.push attribute
+        $generated_uniq_labels_with_value = blacklists
+        [label, values.map { |v| v[:value] }]
+      end
+
+      def random_muf_data(opts = {})
```
Comment (yenkins) [inline spec/lcm/integration/support/user_provisioning_helper.rb]: info: Has approx 9 statements (TooManyStatements) (reek)

File: spec/lcm/integration/support/user_provisioning_helper.rb [hunk 34]
```diff
@@ -0,0 +1,83 @@
...
+              client_id: client.client_id,
+              value: v,
+              label_id: label.identifier,
+              project_id: project.pid
+            }
+          end
+        end.flatten
+      end
+
+      def test_user_filters_brick(options = {})
```
Comment (yenkins) [inline spec/lcm/integration/support/user_provisioning_helper.rb]: info: Has approx 19 statements (TooManyStatements) (reek)

File: spec/lcm/integration/support/user_provisioning_helper.rb [hunk 35]
```diff
@@ -0,0 +1,83 @@
...
+            matching = filters.find { |f| f.related.email == email.downcase }
+            expected_filters.each do |f|
+              value = label.find_value_uri f[:value]
+              spec_env { expect(matching.expression).to include value }
+            end
+          end
+        end
+      end
+
+      def test_users_brick(options = {})
```
Comment (yenkins) [inline spec/lcm/integration/support/user_provisioning_helper.rb]: info: Has approx 8 statements (TooManyStatements) (reek)

File: spec/lcm/integration/support/user_provisioning_helper.rb [hunk 36]
```diff
@@ -0,0 +1,83 @@
+require_relative '../brick_runner'
+
+module Support
+  class UserProvisioningHelper
+    class << self
+      def uniq_label_with_value(project, value_count)
```
Comment (yenkins) [inline spec/lcm/integration/support/user_provisioning_helper.rb]: info: Has approx 14 statements (TooManyStatements) (reek)

File: spec/lcm/integration/support/user_provisioning_helper.rb [hunk 37]
```diff
@@ -0,0 +1,83 @@
...
+      def random_muf_data(opts = {})
+        client = opts[:client]
+        project = opts[:project]
+        users = opts[:users]
+        muf_complexity = opts[:muf_complexity] || 2
+
+        label, values = uniq_label_with_value project, muf_complexity
+        puts "Generated MUF for values #{values} of attribute #{label.title}"
+
+        users.map do |u|
```
Comment (yenkins) [inline spec/lcm/integration/support/user_provisioning_helper.rb]: info: Has the variable name 'u' (UncommunicativeVariableName) (reek)

File: spec/lcm/integration/support/user_provisioning_helper.rb [hunk 38]
```diff
@@ -0,0 +1,83 @@
...
+      end
+
+      def test_user_filters_brick(options = {})
+        test_context = options[:test_context]
+        mufs = options[:mufs]
+        projects = options[:projects]
+
+        BrickRunner.user_filters_brick context: test_context, template_path: '../params/user_filters_brick_e2e.json.erb'
+
+        projects.each do |p|
```
Comment (yenkins) [inline spec/lcm/integration/support/user_provisioning_helper.rb]: info: Has the variable name 'p' (UncommunicativeVariableName) (reek)

File: spec/lcm/integration/support/user_provisioning_helper.rb [hunk 39]
```diff
@@ -0,0 +1,83 @@
...
+      end
+
+      def test_users_brick(options = {})
+        test_context = options[:test_context]
+        projects = options[:projects]
+        user_data = options[:user_data]
+
+        BrickRunner.users_brick context: test_context, template_path: '../params/users_brick_e2e.json.erb'
+
+        projects.each do |p|
```
Comment (yenkins) [inline spec/lcm/integration/support/user_provisioning_helper.rb]: info: Has the variable name 'p' (UncommunicativeVariableName) (reek)

File: spec/lcm/integration/support/user_provisioning_helper.rb [hunk 40]
```diff
@@ -0,0 +1,83 @@
...
+  class UserProvisioningHelper
+    class << self
+      def uniq_label_with_value(project, value_count)
+        blacklists = $generated_uniq_labels_with_value || {}
+        blacklist = blacklists[project.pid] || []
+
+        atts = project.attributes.to_a.sort_by(&:identifier)
+        attribute = atts.shift while !attribute || attribute.date_attribute? || !attribute.primary_label || (blacklist.include? attribute)
+        label = attribute.primary_label
+        sorted_values = label.values.to_a.sort_by { |a| a[:value] }
```
Comment (yenkins) [inline spec/lcm/integration/support/user_provisioning_helper.rb]: info: Has the variable name 'a' (UncommunicativeVariableName) (reek)

**Comment by gdgate** [comment]
Build failed (check pipeline).

- https://checklist.intgdc.com/job/compliance/job/gooddata-ruby-opensource-repository-check/745/ : SUCCESS in 8s
- https://ci.intgdc.com/job/gooddata-ruby-pronto-code-review-zuul/1568/ : FAILURE in 2m 51s
- https://ci.intgdc.com/job/gooddata-ruby-unit-tests-zuul/1829/ : FAILURE in 1m 39s
- https://jenkins-ii.intgdc.com/job/gooddata-ruby/job/gooddata-ruby-check-pipeline/61/ : SUCCESS in 6m 41s

**Comment by gdgate** [comment]
Build failed (check pipeline).

- https://checklist.intgdc.com/job/compliance/job/gooddata-ruby-opensource-repository-check/746/ : SUCCESS in 6s
- https://ci.intgdc.com/job/gooddata-ruby-pronto-code-review-zuul/1569/ : FAILURE in 2m 24s
- https://ci.intgdc.com/job/gooddata-ruby-unit-tests-zuul/1830/ : FAILURE in 3m 08s
- https://jenkins-ii.intgdc.com/job/gooddata-ruby/job/gooddata-ruby-check-pipeline/62/ : SUCCESS in 7m 02s

**Comment by kubamahnert** [comment]
https://travis-ci.org/gooddata/gooddata-ruby/jobs/442165505

**Comment by gdgate** [comment]
Build failed (check pipeline).

- https://checklist.intgdc.com/job/compliance/job/gooddata-ruby-opensource-repository-check/747/ : SUCCESS in 12s
- https://ci.intgdc.com/job/gooddata-ruby-pronto-code-review-zuul/1570/ : FAILURE in 2m 10s
- https://ci.intgdc.com/job/gooddata-ruby-unit-tests-zuul/1831/ : FAILURE in 2m 35s
- https://jenkins-ii.intgdc.com/job/gooddata-ruby/job/gooddata-ruby-check-pipeline/63/ : SUCCESS in 6m 45s

File: lib/gooddata/lcm/actions/synchronize_user_filters.rb [hunk 41]
```diff
@@ -154,40 +158,34 @@ def call(params)
             user_filters.group_by { |u| u[multiple_projects_column] }.flat_map.pmap do |client_id, new_filters|
               users = users_by_project[client_id]
               fail "Client id cannot be empty" if client_id.blank?
+
               c = domain.clients(client_id, data_product)
               if params.segments && !segment_uris.include?(c.segment_uri)
-                puts "Client #{client_id} is outside segments_filter #{params.segments}"
+                params.gdc_logger.warn "Client #{client_id} is outside segments_filter #{params.segments}"
```
Comment (yenkins) [inline lib/gooddata/lcm/actions/synchronize_user_filters.rb]: info: Calls 'params.gdc_logger' 3 times (DuplicateMethodCall) (reek)

Comment (yenkins) [inline lib/gooddata/lcm/actions/synchronize_user_filters.rb]: info: Calls 'params.segments' 4 times (DuplicateMethodCall) (reek)

File: spec/lcm/integration/support/configuration_helper.rb [hunk 42]
```diff
@@ -72,7 +72,7 @@ def ensure_development_project(opts = {})
           helper
         end
       else
-        return create_development_project(opts)
+        create_development_project(opts)
```
Comment (yenkins) [inline spec/lcm/integration/support/configuration_helper.rb]: info: Calls 'create_development_project(opts)' 2 times (DuplicateMethodCall) (reek)

File: spec/lcm/integration/support/user_provisioning_helper.rb [hunk 43]
```diff
@@ -0,0 +1,89 @@
...
+        notempty = sorted_values.reject { |v| v[:value].empty? }
+        values = notempty.slice(0, value_count)
+
+        blacklists[project.pid] = blacklist.push attribute
+        $generated_uniq_labels_with_value = blacklists
+        [label, values.map { |v| v[:value] }]
+      end
+
+      def muf_data(opts = {})
```
Comment (yenkins) [inline spec/lcm/integration/support/user_provisioning_helper.rb]: info: Has approx 9 statements (TooManyStatements) (reek)

**Comment by gdgate** [comment]
Build failed (check pipeline).

- https://checklist.intgdc.com/job/compliance/job/gooddata-ruby-opensource-repository-check/763/ : SUCCESS in 7s
- https://ci.intgdc.com/job/gooddata-ruby-pronto-code-review-zuul/1585/ : FAILURE in 2m 39s
- https://ci.intgdc.com/job/gooddata-ruby-unit-tests-zuul/1846/ : FAILURE in 1m 16s
- https://jenkins-ii.intgdc.com/job/gooddata-ruby/job/gooddata-ruby-build-pipeline/25/ : SUCCESS in 8m 06s

**Comment by kubamahnert** [comment]
https://travis-ci.org/kubamahnert/gooddata-ruby/builds/449361143 pronto is not passing, but I'd vote to ignore it as the mass is quite low.

**Comment by kubamahnert** [comment]
https://travis-ci.org/kubamahnert/gooddata-ruby/builds/449366978 TMA-1152 is failing

**Comment by kubamahnert** [comment]
ok to test

**Comment by gdgate** [comment]
Build failed (check pipeline).

- https://checklist.intgdc.com/job/compliance/job/gooddata-ruby-opensource-repository-check/766/ : SUCCESS in 10s
- https://ci.intgdc.com/job/gooddata-ruby-pronto-code-review-zuul/1588/ : FAILURE in 1m 11s
- https://ci.intgdc.com/job/gooddata-ruby-unit-tests-zuul/1849/ : FAILURE in 1m 06s
- https://jenkins-ii.intgdc.com/job/gooddata-ruby/job/gooddata-ruby-build-pipeline/28/ : SUCCESS in 6m 45s

**Pull request merged**